xwiki-platform-core/xwiki-platform-messagestream/xwiki-platform-messagestream-api/src/main/java/org/xwiki/messagestream/internal/AbstractMessageStreamNotificationFilter.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.messagestream.internal;
+
+import java.util.Collection;
+import java.util.Objects;
+
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.notifications.NotificationFormat;
+import org.xwiki.notifications.filters.NotificationFilter;
+import org.xwiki.notifications.filters.NotificationFilterPreference;
+import org.xwiki.notifications.filters.NotificationFilterType;
+import org.xwiki.notifications.filters.expression.ExpressionNode;
+import org.xwiki.notifications.preferences.NotificationPreference;
+import org.xwiki.notifications.preferences.NotificationPreferenceProperty;
+
+/**
+ * Base class for the message notification filters.
+ *
+ * @version $Id$
+ * @since 12.9RC1
+ */
+public abstract class AbstractMessageStreamNotificationFilter implements NotificationFilter
+{
+    @Override
+    public boolean matchesPreference(NotificationPreference preference)
+    {
+        return Objects.equals(getEventType(), getEventType(preference));
+    }
+
+    @Override
+    public ExpressionNode filterExpression(DocumentReference user,
+        Collection<NotificationFilterPreference> filterPreferences, NotificationFilterType type,
+        NotificationFormat format)
+    {
+        return null;
+    }
+
+    @Override
+    public ExpressionNode filterExpression(DocumentReference user,
+        Collection<NotificationFilterPreference> filterPreferences, NotificationPreference preference)
+    {
+        return null;
```

Comment:
I'm not sure that's good: I can see that `DirectMessageStreamNotificationFilter` is actually overriding it, I'm wondering why it's not the case in the others. And why not on the other filterExpression method. I don't know much this API so not sure when those methods are actually used.

Reply:
IIUC, this contributes to the store filtering when retrieving the message send to an user activity stream.
But AFAICS, the messages are already send only to the relevant recipients, so as long a we do not provide more specific message stream preferences, retuning null (ie, no filter) is fine.
And AFAIU (but I might be missing something), the `DirectMessageStreamNotificationFilter` filter is a tautology and I could not observe different results by simply replacing the body with `return null` here too.

Reply:
> But AFAICS, the messages are already send only to the relevant recipients, so as long a we do not provide more specific message stream preferences, retuning null (ie, no filter) is fine.

Be careful, maybe it's not used anymore after the changes performed by @tmortagne to use pre-filtering, but it's certainly used when you disable the pre-filtering. You should make a manual test by disabling it in xwiki.properties and ensure it's still working properly.

Reply:
I have been using a similar construct for a custom filter some time ago for an 11.x XWiki instance. Returning ``null`` here is fine, and handles by the calling side.

Reply:
I've tested without pre-filtering and couldn't find an issue. In addition, my changes are pure refactoring and does not modify the implementations of the existing filter. The most common methods are moved to the abstract method and only personal override `filterExpression`.
If something is wrong with how things are currently working, it should probably be addresses in another issue.

Reply:
So from what I see there's 2 different cases: 
`NotificationFilter#filterExpression(DocumentReference, Collection<NotificationFilterPreference>, NotificationFilterType, NotificationFormat)` (not that one, but the one above) should be ok to return null since it's only use for inclusive/exclusive filters: basically it's used in ScopeNotificationFilter only. 

Now this method we're discussing here is used when computing the query in `QueryExpressionGenerator`: when you don't override this filter, calling `EventSearcher#searchStoreEvents` will return wrong results, it will also include some more events that shouldn't be when you don't use pre-filtering.
You see it working in XWiki because we actually do not only rely on `EventSearcher#searchStoreEvents` but we also call the `NotificationFilter#filterEvent` method in `DefaultParametrizedNotificationManager` after getting the results. 

IMO as part of your refactoring you should provide clean filters with this method properly overridden in each classes, so that we wouldn't get more results than needed from the query when not using pre-filtering.

Reply:
Note that I think we are doing this double filtering (from query first, and then from filterEvent) because some notifications couldn't be filtered only from the query (can you confirm @gdelhumeau or @tmortagne ?). Now it looks like those filters you're working on can.

Reply:
The filtering is now realized fully in `filterEvent`, and nothing more is needed in the `filterExpression` method that are now all retuning `null` for the message stream filters.